pkg/cloud/aliyunprovider.go

```diff
@@ -0,0 +1,828 @@
+		err := aliyun.loadAliyunAuthSecretAndSetEnv(true)
+		if err != nil {
+			return nil, fmt.Errorf("unable to set the aliyun key/secret from config file")
+		}
+		// set custom pricing keys too
+		config.AliyunServiceKeyName = env.GetAliyunAccessKeyID()
+		config.AliyunServiceKeySecret = env.GetAliyunAccessKeySecret()
+	}
+
+	if config.AliyunServiceKeyName == "" && config.AliyunServiceKeySecret == "" {
+		return nil, fmt.Errorf("failed to get the access key for the current alibaba account")
+	}
+
+	aliyun.accessKey = &credentials.AccessKeyCredential{AccessKeyId: env.GetAliyunAccessKeyID(), AccessKeySecret: env.GetAliyunAccessKeySecret()}
+	aliyun.loadedAccessKey = true
+
+	return aliyun.accessKey, nil
+}
+
+func (aliyun *Aliyun) DownloadPricingData() error {
+	aliyun.DownloadPricingDataLock.Lock()
+	defer aliyun.DownloadPricingDataLock.Unlock()
+
+	var aak *credentials.AccessKeyCredential
+	var err error
+
+	if !aliyun.loadedAccessKey {
+		aak, err = aliyun.GetAliyunAccessKey()
+		if err != nil {
+			log.Errorf("Unable to get the access key information")
+			return fmt.Errorf("unable to get the access key information")
+		}
+	} else {
+		aak = aliyun.accessKey
+	}
+
+	c, err := aliyun.Config.GetCustomPricingData()
+	if err != nil {
+		log.Errorf("Error downloading default pricing data: %s", err.Error())
+		return fmt.Errorf("error downloading default pricing data: %s", err.Error())
+	}
+
+	// // Get the nodes in Aliyun provider, Once alibaba cloud is setup
+	// // and populate data from the node object to resemble the data in hardcodeNodes
+	nodeList := aliyun.Clientset.GetAllNodes()
+
+	var client *sdk.Client
+	var signer *signers.AccessKeySigner
+	var ok bool
+	var pricingObj *AliyunPricing
+	var lookupKey string
+	aliyun.clients = make(map[string]*sdk.Client)
+	aliyun.Pricing = make(map[string]*AliyunPricing)
+
+	// TO-DO: Add disk price adjustment by parsing the local disk information and putting it as a param in describe Price function.
+	for _, node := range nodeList {
+		slimK8sNode := generateSlimK8sNodeFromV1Node(node)
+		lookupKey, err = determineKeyForPricing(slimK8sNode)
+		if _, ok := aliyun.Pricing[lookupKey]; ok {
+			log.Infof("Pricing information for node with same features %s already exists hence skipping", lookupKey)
+			continue
+		}
+		// TO-DO: Check if Node pricing already available , skip it if available
+		if client, ok = aliyun.clients[slimK8sNode.RegionID]; !ok {
+			client, err = sdk.NewClientWithAccessKey(slimK8sNode.RegionID, aak.AccessKeyId, aak.AccessKeySecret)
+			if err != nil {
+				return fmt.Errorf("access key provided does not have access to location %s", slimK8sNode.RegionID)
+			}
+			aliyun.clients[slimK8sNode.RegionID] = client
+		}
+		signer = signers.NewAccessKeySigner(aak)
+		pricingObj, err = processDescribePriceAndCreateAliyunPricing(client, slimK8sNode, signer, c)
+
+		if err != nil {
+			return err
+		}
```

Comment:
Would you please wrap this error return with a helpful description of what was being attempted?